docs/pages/pmd/userdocs/extending/writing_rules_intro.md

```diff
@@ -0,0 +1,141 @@
+---
+title: Introduction to writing PMD rules
+tags: [extending, userdocs, getting_started]
+summary: "Writing your own PMD rules TODO"
+last_updated: July 2018 (6.6.0)
+permalink: pmd_userdocs_extending_writing_rules_intro.html
+author: Clément Fournier <clement.fournier76@gmail.com>
+---
+
+PMD is a framework to perform code analysis. You can create your own rules to
+check for patterns specific to your codebase, or the coding practices of your
+team.
+
+## How rules work: the AST
+
+Before running rules, PMD parses the source file into a data structure called an
+**abstract syntax tree (AST)**. This tree represents the syntactic structure of the
+code, and encodes syntactic relations between source code elements. For instance,
+in Java, method declarations belong to a class: in the AST, the nodes representing
+method declarations will be descendants of a node representing the declaration of
+their enclosing class. This representation is thus much richer than the original
+source code (which, for a program, is just a chain of characters), or the token
+chain produced by a lexer (which is e.g. what Checkstyle works on). For example:
+
+<table>
+<colgroup>
+<col width="40%" />
+<col width="70%" />
+</colgroup>
+<thead>
+<tr class="header">
+<th>Sample code (Java)</th>
+<th>AST</th>
+</tr>
+</thead>
+<tbody>
+<tr>
+<td markdown="block">
+
+```java
+class Foo extends Object {
+
+}
+```
+
+</td>
+<td markdown="block">
+
+```java
+└─ CompilationUnit
```

Comment:
Nice styling :+1:

Reply:
This is generated by the designer :) You can choose to export a subtree to text